### PR TITLE
DBZ-6964 Debezium jdbc sink create table failure with Duplicate column name '_geometry'

### PR DESCRIPTION
**Pull request by harveyyue**
https://issues.redhat.com/browse/DBZ-6964

### CODE REVIEW

**Comment by jpechane** [comment]
@harveyyue Before applying this PR I'd like to clarify something. Do I inderstand it correctly that in this case there is only one instance of geometry schema that is shared by all columns and thus the distinc schema parameters per-column are lost?

**Comment by harveyyue** [comment]
> @harveyyue Before applying this PR I'd like to clarify something. Do I inderstand it correctly that in this case there is only one instance of geometry schema that is shared by all columns and thus the distinc schema parameters per-column are lost?

@jpechane Yes, please see the avro schema fragment as below, I believe that avro converter always reuse the same record schema "Geometry" and ignore the other column's schema paramters
`{
				"name": "_geometry",
				"type": ["null", {
					"type": "record",
					"name": "Geometry",
					"namespace": "io.debezium.data.geometry",
					"fields": [{
						"name": "wkb",
						"type": "bytes"
					}, {
						"name": "srid",
						"type": ["null", "int"],
						"default": null
					}],
					"connect.doc": "Geometry",
					"connect.version": 1,
					"connect.parameters": {
						"__debezium.source.column.type": "GEOMETRY",
						"__debezium.source.column.name": "_geometry"
					},
					"connect.name": "io.debezium.data.geometry.Geometry"
				}],
				"default": null
			}, {
				"name": "_linestring",
				"type": ["null", "io.debezium.data.geometry.Geometry"],
				"default": null
			}, {
				"name": "_polygon",
				"type": ["null", "io.debezium.data.geometry.Geometry"],
				"default": null
			}`

**Comment by jpechane** [comment]
@harveyyue Then IMHO this is a band-aid and not real fix. IMHO we need to find out if this behaviour is correct or not in the sense of Avro converter. Are two schemas with different parameters two different schemas? It is important to check it with Avro Converter guys. If not then we have much more serious problem as a lot of Debezium functionality relies on schema parameters and this would break a lot of things.
OTOH if these would not be two schemas then I'd miss the prupose of parameters.
If it is idenitifed as Avro Converter bug then I'd be fine with committing the fix as temporrary solution with scheduling for removal - and probably I'd even guard it with an internal config option so it is opt-in and can be easily removed.

**Comment by mfvitale** [comment]
Just for completeness ( even if it's not strictly related) I encountered similar issue with this https://issues.redhat.com/browse/DBZ-6836 and I have opened an issue to [confluent](https://github.com/confluentinc/schema-registry/issues/2740) and [apicurio](https://github.com/Apicurio/apicurio-registry/issues/3606).

In brief, they uses a cache to indentify cycles in structs and this cache have only the name of the struct as key. If you have same structure type with different parameters, only the ones from the first element inserted in the cache will be used..

**Comment by jpechane** [comment]
OK, then the question is - is the schema parameter concept even compatible with Avro at all?

@Naros This might be of some interest for you.

**Comment by Naros** [comment]
Hi @harveyyue, in discussion with @jpechane, he made a neat suggestion here that we should likely explore as a way forward, as I think there is no way to solve this quickly with Avro and Kafka Connect.

The idea is to introduce a new single message transformation that is responsible for ingesting the schema parameters for each field within a Debezium event, and it adds a new payload top-level info block to the payload that the JDBC sink could consume these parameters from as a alternative.  

On the sink side, if it's a Debezium event, we would prioritize the schema parameters section as we do now. If that area has no details, we would then check for this schema parameters information block, and if that exists, source the parameters from that instead.  If that block doesn't exist, it would act like it does presently with a failure.

Would you like to contribute this idea as the fix for this use case?

**Comment by harveyyue** [comment]
@Naros Ok, I will take this new SMTp

**Comment by harveyyue** [comment]
The sibling pr: https://github.com/debezium/debezium/pull/5037

File: src/main/java/io/debezium/connector/jdbc/SinkRecordDescriptor.java [hunk 1]
```diff
@@ -177,7 +177,8 @@ private FieldDescriptor(Schema schema, String name, boolean key, DatabaseDialect
             this.typeName = type.getTypeName(dialect, schema, key);
 
             this.name = name;
-            this.columnName = SchemaUtils.getSourceColumnName(schema).orElse(name);
+            // Avoid avro schema reuse of Geometry record structure to get same column name
+            this.columnName = isStructType() ? name : SchemaUtils.getSourceColumnName(schema).orElse(name);
```
Comment (jpechane) [inline src/main/java/io/debezium/connector/jdbc/SinkRecordDescriptor.java:181]: @harveyyue Should this be limited to struct only? IIUC any type that has metadata in parameters like Decimal will be affected.

Comment (harveyyue) [reply src/main/java/io/debezium/connector/jdbc/SinkRecordDescriptor.java:181]: > @harveyyue Should this be limited to struct only? IIUC any type that has metadata in parameters like Decimal will be affected.

Yes, only struct type, primitive type is work fine.

Comment (jpechane) [reply src/main/java/io/debezium/connector/jdbc/SinkRecordDescriptor.java:181]: @harveyyue The problem as descibed above is that if there are two arbitrary types differing with schema parameters only then only one (probably the first) is used and the rest is forgotten. In that case I believe the schema parameters for all types should be used. Also IMHO the problem with schema paramaters is thus not limited to this place but might be present elsewhere too. WDYT?

Comment (harveyyue) [reply src/main/java/io/debezium/connector/jdbc/SinkRecordDescriptor.java:181]: > @harveyyue The problem as descibed above is that if there are two arbitrary types differing with schema parameters only then only one (probably the first) is used and the rest is forgotten. In that case I believe the schema parameters for all types should be used. Also IMHO the problem with schema paramaters is thus not limited to this place but might be present elsewhere too. WDYT?

Yes, I think this case is limited when using avro, protobuf or other format converter scenario.

Comment (jpechane) [reply src/main/java/io/debezium/connector/jdbc/SinkRecordDescriptor.java:181]: @harveyyue I merged the SMT PR. To make sure the issue is really addresed could you please prepare an integration test running for MySQL with Apicurio and Avro that will have two geometry columns and two decimal columns with different constraints and check that everything is properly proagated?
That way we'll be 100 % sure that the general issue was nailed.
Thanks

**Comment by rk3rn3r** [comment]
Based on today's call with @jpechane I am going to look into this one.

**Comment by Naros** [comment]
Given the discussion here in this PR and that we've moved the connector the core repository, I'm going to proceed with closing this PR.  Please continue on the Jira issue and open a new PR in the core repository if any further changes are needed.

**Pull request closed without merging**